### PR TITLE
Common base class for Parameters,Calculations,Visibilities

### PR DESCRIPTION
**Pull request by gerw**
This patch addresses #122. It also provides some simple tests for these classes.

Please let me know if you have a better name for the base class.

### CODE REVIEW

**Comment by gerw** [comment]
I think `pytest` fails due to a problem with permissions, see https://github.com/MishaKav/pytest-coverage-comment/issues/30.

It might be fixed by adding
```
permissions:
  pull-requests: write
```
to `.github/workflows/pytest.yml`? Or does this problem only happens for pull request from forks?

**Comment by Bouni** [comment]
Strange, for me it worked when I re-activated this feature 🤔

**Comment by Bouni** [comment]
@gerw please push a CI trigger commit to test #134

**Comment by gerw** [comment]
@Bouni: Just triggering the CI did not help.

After I rebased to `main`, it worked.

File: tests/test_visibilities.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+from luxtronik.visibilities import Visibilities
+
+
+class TestVisibilities:
+    """Test suite for Visibilities"""
+
+    def test_init(self):
+        """Test cases for initialization"""
+        visibilities = Visibilities()
+        assert visibilities.name == "Visibility"
```
Comment (kbabioch) [inline tests/test_visibilities.py:14]: This test might be missing for `Parameter`, as I can only see it for `Visibility` and `Calculation`.

**Comment by gerw** [comment]
I just realized that `pytest` in the CI did not recognized my new test files. Does anybody know why?

File: luxtronik/data_vector.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+"""Provides a base class for parameters, calculations, visibilities."""
+import logging
+
+from luxtronik.datatypes import Unknown
+
+
+class DataVector:
```
[the remark this answers is not in the document]
Comment (gerw) [reply luxtronik/data_vector.py:7]: Because the `_data` inside is a tuple of values.

File: tests/test_visibilities.py (hunk 1 above)
Comment (gerw) [reply tests/test_visibilities.py:14]: Added.

File: luxtronik/data_vector.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            if entry is not None:
+                entry.raw = data
+            else:
+                # self.logger.warning(f"Entry '%d' not in list of {self.name}", index)
+                entry = Unknown(f"Unknown_{self.name}_{index}")
+                entry.raw = data
+                self._data[index] = entry
+
+    def _lookup(self, target, with_index=False):
+        """Lookup entry by either id or name."""
```
[the remark this answers is not in the document]
Comment (gerw) [reply luxtronik/data_vector.py]: Added.

File: luxtronik/data_vector.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+    logger = logging.getLogger("Luxtronik.DataVector")
+    name = "DataVector"
+
+    def __init__(self):
+        """Initialize DataVector class."""
+        self._data = {}
+
+    def __iter__(self):
+        return iter(self._data.items())
```
[the remark this answers is not in the document]
Comment (gerw) [reply luxtronik/data_vector.py]: Added.

**Comment by kbabioch** [comment]
> I just realized that `pytest` in the CI did not recognized my new test files. Does anybody know why?

No, don't quite understand this (yet). Just checked out your branch locally and invoked the same command on my machine that is executed in the pipeline. For me it looks like this:

```
PYTHONPATH=. pytest --junitxml=pytest.xml --cov-report=term-missing:skip-covered --cov=luxtronik tests/ | tee pytest-coverage.txt
============================= test session starts ==============================
platform linux -- Python 3.11.3, pytest-7.4.0, pluggy-1.0.0
rootdir: /home/kbabioch/alpha-innotec-hacking/python-luxtronik
plugins: cov-4.1.0, anyio-3.7.0
collected 110 items

tests/test_calculations.py .                                             [  0%]
tests/test_datatypes.py ...ssss......................................... [ 44%]
......................................................                   [ 93%]
tests/test_parameters.py ......                                          [ 99%]
tests/test_visibilities.py .                                             [100%]

- generated xml file: /home/kbabioch/alpha-innotec-hacking/python-luxtronik/pytest.xml -

---------- coverage: platform linux, python 3.11.3-final-0 -----------
Name                                  Stmts   Miss  Cover   Missing
-------------------------------------------------------------------
luxtronik/__init__.py                   134    106    21%   38-54, 61-66, 69-73, 79, 83, 102-113, 116-119, 122-142, 145-160, 163-180, 183-198, 202-204, 208-209, 213-214
luxtronik/__main__.py                    21     21     0%   2-48
luxtronik/datatypes.py                  235     12    95%   37, 42, 47, 57, 72-75, 80-83, 92
luxtronik/discover.py                    43     34    21%   25-77
luxtronik/scripts/dump_changes.py        44     44     0%   5-93
luxtronik/scripts/dump_luxtronik.py      28     28     0%   5-64
-------------------------------------------------------------------
TOTAL                                   588    245    58%

6 files skipped due to complete coverage.

======================== 106 passed, 4 skipped in 0.34s ========================
```

So the new test classes have been picked up. Not sure why this is not the case here in this pull request. Maybe some kind of caching? Re-triggering the job did not help, though, will have to investigate further.

**Review by kbabioch — APPROVED** [review]

**Comment by kbabioch** [comment]
Overall the changes look fine to me. Unit tests also seem to work fine (locally), so in case you want to go ahead with the merge, it's fine with me. The unit test issue needs to be addressed eventually, though.

**Pull request merged**